## Cancer genomes tolerate deleterious coding mutations through somatic copy-number amplifications of wild-type regions

[comment]: <> (Replace with the correct DOI)
[![](https://img.shields.io/badge/doi-10.1101/2021.02.13.429885-rec.svg)](https://doi.org/10.1101/2021.03.13.429885)

### Abstract
make it short!

Cancers evolve under the accumulation of thousands of somatic mutations and chromosomal aberrations. Most protein-coding genes lack detectable signals of selection suggesting that most mutations are neutral events with no effects on cancer fitness. Conversely, in silico evolutionary simulations, in vitro and in vivo proliferation assays and whole-genome CRISPR-based knock-out screens within tumor cell lines demonstrated the deleterious effects of most mutations and gene KO (reduced in vitro proliferation and in vivo tumor growth).  This raises the question of how tumors tolerate such large amounts of deleterious mutations and why we cannot detect strong signals of negative selection acting on cancer-essential genes. Using 8,690 primary tumor samples from The Cancer Genome Atlas (TCGA) dataset, we performed an unbiased relationship analysis between somatic amplification and mutation frequency under different conditions and cancer types. We demonstrated that copy-number amplifications frequently cover mutation prone regions, increasing genomic tolerance towards the deleterious impact of mutations by creating extra safe copies of wild-type regions, eventually protecting the genes. We showed that these potential compensatory events are highly influenced by gene and mutation properties, with haploinsufficient genes show a higher tendency to be protected from mutations via amplifications. Moreover, over-representation analysis of unprotected gene sets revealed unessential and cancer-unrelated function enrichment; conversely, protected gene sets contain cellular essential and cancer-associated functions. A deeper understanding of these tumor genome-shaping compensatory events could reveal new cancer vulnerabilities and might guide novel therapeutic intervention.


### Reproduce analysis and figures


#### (1) Clone this repository

The first step is to clone this repository (you need git, if you haven't installed it, check [this link](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)):
```bash
git clone https://github.com/fabio-alfieri/mutation_compensation.git
```
#### (2) Dowload supplementary data from Zenodo 

Download data.zip from [Zenodo](https://doi.org/10.5281/zenodo.7065200), put it in the cloned GitHub folder and unzip it.

#### (3) Dowload supplementary data from Zenodo 

If not, install [Conda]
```bash
```
